lib/propolis/src/hw/virtio/pci.rs

```diff
@@ -395,6 +410,22 @@ impl PciVirtioState {
             state.status = val;
         }
     }
+
+    /// Set the VirtIO device in a Failed state.
+    ///
+    /// This will require the guest OS to reset the device before it can use it
+    /// for anything further.
+    fn set_failed(
+        &self,
+        _dev: &dyn VirtioDevice,
+        state: &mut MutexGuard<VirtioState>,
+    ) {
+        if !state.status.contains(Status::FAILED) {
+            state.status.insert(Status::FAILED);
+            // XXX: interrupt needed?
```

Comment:
I read [section 2.1 of the Virtio 1.2 spec](http://docs.oasis-open.org/virtio/virtio/v1.0/cs04/virtio-v1.0-cs04.html#x1-100001) (versions 1.0 and 1.1 have the same language in this section) and came away thinking that

- This should set `DEVICE_NEEDS_RESET` and not `FAILED` (IIUC the latter is for guest drivers to set and the former is for devices to set, even if device failed while responding to a driver write), and
- If that's correct and the `DRIVER_OK` bit was previously set, then a configuration interrupt is required after doing this.

Reply:
I'll fix this to use `NEEDS_RESET` rather than `FAILED`.  As for the configuration interrupt work, I'll open an issue to deal with that separately.